src/errors.rs

```diff
@@ -14,35 +17,59 @@ pub fn handle_panic(info: &PanicInfo<'_>, send_app_instructions: &SyncSender<App
         None => info.payload().downcast_ref::<String>().map(|s| &**s),
     };
 
+    let mut err_ctx: ErrorContext = OPENCALLS.with(|ctx| ctx.borrow().clone());
+
     let backtrace = match (info.location(), msg) {
         (Some(location), Some(msg)) => format!(
-            "\nthread '{}' panicked at '{}': {}:{}\n{:?}",
+            "thread '{}' panicked at '{}': {}:{}\n{:#?}\n{:?}",
             thread,
             msg,
             location.file(),
             location.line(),
+            err_ctx,
             backtrace
         ),
         (Some(location), None) => format!(
-            "\nthread '{}' panicked: {}:{}\n{:?}",
+            "thread '{}' panicked: {}:{}\n{:#?}\n{:?}",
             thread,
             location.file(),
             location.line(),
+            err_ctx,
             backtrace
         ),
         (None, Some(msg)) => format!(
-            "\nthread '{}' panicked at '{}'\n{:?}",
-            thread, msg, backtrace
+            "thread '{}' panicked at '{}'\n{:#?}\n{:?}",
+            thread, msg, err_ctx, backtrace
+        ),
+        (None, None) => format!(
+            "thread '{}' panicked\n{:#?}\n{:?}",
+            thread, err_ctx, backtrace
         ),
-        (None, None) => format!("\nthread '{}' panicked\n{:?}", thread, backtrace),
     };
 
     if thread == "main" {
         println!("{}", backtrace);
         process::exit(1);
     } else {
+        err_ctx.add_call("panic_hook(handle_panic)");
         send_app_instructions
-            .send(AppInstruction::Error(backtrace))
+            .send((err_ctx, AppInstruction::Error(backtrace)))
             .unwrap();
     }
 }
+
+#[derive(Clone, Debug)]
+pub struct ErrorContext {
+    calls: Vec<String>,
```

Comment:
A few things about this:
1. Can we change `String` here to work with the instruction types we already have? eg. something like `enum { ScreenInstruction, PtyInstruction, AppInstruction }` - I think that will both help performance and prevent overzealous developers from adding weird stuff to this in the future :)
2. Speaking about performance, while I definitely understand why it makes sense for this to be a `Vec` - I think in our case it might be a bit of an issue. Ideally, I would like this type to be `Copy`, because we would be copying it around A LOT (every event we get from the pty, for example, which is every character print, every style change, etc. if we're printing 10000 lines to the terminal, we will really feel this). And of course, we can't do this with a vec... So I was thinking maybe make this a fixed size array. We'll have it as a separate const type (maybe call it something like `MAX_THREAD_CALL_STACK`), we'll set its length to the maximum we have now (I guess it's 3? 4?) and in the future we can make it larger if we need to. What do you think?

Reply:
I was in fact thinking about the second point myself and was looking at the [arrayvec](https://docs.rs/arrayvec/0.5.2/arrayvec/) crate for the same.

If I understand this correctly, you're suggesting-
```rust
pub enum ContextType {
    ScreenInst(ScreenInstruction),
    PtyInst(PtyInstruction),
    AppInst(AppInstruction),
    IPCServer,
    StdinHandler,
    AsyncTask,
}
```
and have a fixed size array inside `ErrorContext` so that the elements are `Copy` (or basically stored on the stack itself).
I see 2 things ahead of us-
1. We'll have to custom implement Debug for the `***Instruction` enums to prevent enum variants with `Vec` types (such as `ScreenInstruction::ApplyLayout((Layout, Vec<RawFd>))` from filling up the terminal.
2. For `ErrorContext` to implement `Copy` trait, we'll need to have `ContextType` and all the `***Instruction` enums to implement `Copy` trait. These enums have variants with `Vec` type data which won't allow us to have them stored on stack. I maybe misunderstanding your idea of the enum variants of `ContextType`, so please correct me if I'm wrong.

Reply:
You understood me perfectly, and are actually one step ahead of me. :)
I forgot about the `Vec`s in the events, and actually now noticed an old TODO of mine to fix them: https://github.com/mosaic-org/mosaic/blob/main/src/pty_bus.rs#L65 :)

I have a suggestion until we fix those instructions (because they should definitely be `Copy` too):
Let's make `ContextType` like this:
```rust
pub enum ContextType {
    Screen(ScreenContext),
    Pty(PtyContext),
    App(AppContext),
    IPCServer,
    StdinHandler,
    AsyncTask,
}
```
And then we can have for example:
```rust
pub enum ScreenContext {
    Pty,
    Render,
    NewPane,
    // ...
}

impl From<ScreenInstruction> for ScreenContext {
    fn from(screen_instruction: ScreenInstruction) -> Self {
        match ScreenInstruction {
            ScreenInstruction::Pty(_) => ScreenContext::Pty,
            // ...
        }
    }
}
```

What do you think?

Reply:
Cool. Sounds good to me. What do you think about the arrayvec crate?

Reply:
I trust your judgment.

Reply:
So, I found out that `ArrayVec` doesn't implement `Copy` (for reasons mentioned [here](https://github.com/bluss/arrayvec/issues/32)) so we need to explicitly call `.clone()` but it definitely facilitates stack allocation ensuring better performance.

Reply:
Wouldn't that be defeating the purpose though? Why not use an array with a fixed length and then we can also implement Copy?

Reply:
With arrays, we'll have to initialize it. So should we have another enum variant `ContextType::Empty` for that? So `add_call()` will replace the first `Empty` entry it encounters with the current thread instruction.

Reply:
Either that or make it an `Option<ContextType>`, whatever feels better for you.

Reply:
`Option` will add 8 bytes to the current size of `ContextType`. So, I'll go with `Empty`.

Reply:
Done!